### PR TITLE
Allow alternative SSL libaries with a warning

### PR DESCRIPTION
**Pull request by pquentin**
Closes https://github.com/urllib3/urllib3/issues/3020

I'll have to run those two tests manually since the test suite does not cover them:

- [x] LibreSSL
- [x] OpenSSL 1.0.2

### CODE REVIEW

**Review by sigmavirus24 — APPROVED** [review]

File: src/urllib3/__init__.py [hunk 1]
```diff
@@ -30,17 +30,19 @@
...
 else:
-    # fmt: off
-    if (
-        not ssl.OPENSSL_VERSION.startswith("OpenSSL ")
-        or ssl.OPENSSL_VERSION_INFO < (1, 1, 1)
-    ):  # Defensive:
+    if not ssl.OPENSSL_VERSION.startswith("OpenSSL "):  # Defensive
+        warnings.warn(
+            "urllib3 v2.0 only supports OpenSSL 1.1.1+, currently "
+            f"the 'ssl' module is compiled with {ssl.OPENSSL_VERSION}. "
```
Comment (sethmlarson) [inline src/urllib3/__init__.py:36]: Let's add a `!r` suffix to this format

**Review by sethmlarson — APPROVED** [review]
This is great, I only had one comment. Could you share which configuration you used to test LibreSSL and OpenSSL 1.0.2?

File: src/urllib3/__init__.py (hunk 1 above)
Comment (pquentin) [reply src/urllib3/__init__.py]: Done in [`47cce87` (#3024)](https://github.com/urllib3/urllib3/pull/3024/commits/47cce8742b4c5722f6bde4c5de984fcfdfb67f4c) for both messages

**Comment by pquentin** [comment]
This is ready for another round of review. I'd be happy to update the wording of the message and/or the changelog entry.

File: src/urllib3/__init__.py [hunk 2]
```diff
@@ -30,17 +30,19 @@
 except ImportError:
     pass
 else:
-    # fmt: off
-    if (
-        not ssl.OPENSSL_VERSION.startswith("OpenSSL ")
-        or ssl.OPENSSL_VERSION_INFO < (1, 1, 1)
-    ):  # Defensive:
+    if not ssl.OPENSSL_VERSION.startswith("OpenSSL "):  # Defensive:
```
Comment (pquentin) [inline src/urllib3/__init__.py:33]: This allows things that are neither OpenSSL nor LibreSSL, which I think is OK but you can disagree.

File: src/urllib3/__init__.py [hunk 3]
```diff
@@ -30,17 +30,19 @@
...
+            "urllib3 v2.0 only supports OpenSSL 1.1.1+, currently "
+            f"the 'ssl' module is compiled with {ssl.OPENSSL_VERSION!r}. "
+            "See: https://github.com/urllib3/urllib3/issues/3020",
+            exceptions.NotOpenSSLWarning,
+        )
+    elif ssl.OPENSSL_VERSION_INFO < (1, 1, 1):  # Defensive:
         raise ImportError(
             "urllib3 v2.0 only supports OpenSSL 1.1.1+, currently "
-            f"the 'ssl' module is compiled with {ssl.OPENSSL_VERSION}. "
+            f"the 'ssl' module is compiled with {ssl.OPENSSL_VERSION!r}. "
```
Comment (illia-v) [inline src/urllib3/__init__.py:43]: A little nitpick ‒ an error message in "Common upgrading issues" needs quotes around "OpenSSL 1.0.2k-fips  26 Jan 2017" after this change 🙂
https://github.com/urllib3/urllib3/blob/de9caa83ed6dc1ad26d78329fa4298131aed46e2/docs/v2-migration-guide.rst?plain=1#L190

Comment (pquentin) [reply src/urllib3/__init__.py:43]: Thanks, done in [`4e26d7f` (#3024)](https://github.com/urllib3/urllib3/pull/3024/commits/4e26d7f7dd5438379c4bb3fbbde113282a703603)

**Review by illia-v — APPROVED** [review]
@pquentin thanks for working on making the upgrade to 2.0 more convenient!

**Comment by pquentin** [comment]
I tried running the test suite with LibreSSL but I'm currently stuck at just installing cryptography.

**Comment by sigmavirus24** [comment]
@pquentin maybe reach out to one of the Gentoo folks on the other issue to get their help?

**Review by sethmlarson — APPROVED** [review]
I'm happy with the change, but I want confirmation that our test suite passes for both platforms noted above.

**Comment by pquentin** [comment]
@stefan11111 @orbea Can I get one of you interested in running the test suite under LibreSSL? I think we have instructions in our docs but essentially it's `pip install nox` then `nox -s test-3.11`.

Thanks!

**Comment by stefan11111** [comment]
@pquentin 
On gentoo we do not use pip, instead we install all python packages through portage, the system package manager.
```
$ nox -s test-3.11
nox > Failed to load Noxfile /home/stefan/noxfile.py, no such file exists.
```

**Comment by pquentin** [comment]
Thanks. Can you please do the same thing but in the urllib3 directory after having cloned that pull request?

**Comment by stefan11111** [comment]
```
~/Downloads/urllib3 $ nox -s test-3.11
nox > Running session test-3.11
nox > Creating virtual environment (virtualenv) using python3.11 in .nox/test-3-11
nox > python -m pip install -r dev-requirements.txt
nox > python -m pip install '.[socks,secure,brotli,zstd]'
nox > pip --version
pip 23.1.2 from /home/stefan/Downloads/urllib3/.nox/test-3-11/lib/python3.11/site-packages/pip (python 3.11)
nox > python --version
Python 3.11.3
nox > python -c 'import struct; print(struct.calcsize('"'"'P'"'"') * 8)'
64
nox > python -m OpenSSL.debug
pyOpenSSL: 23.1.1
cryptography: 40.0.2
cffi: 1.15.1
cryptography's compiled against OpenSSL: OpenSSL 3.1.0 14 Mar 2023
cryptography's linked OpenSSL: OpenSSL 3.1.0 14 Mar 2023
Python's OpenSSL: LibreSSL 3.7.2
Python executable: /home/stefan/Downloads/urllib3/.nox/test-3-11/bin/python
Python version: 3.11.3 (main, May  8 2023, 01:46:20) [GCC 13.1.0]
Platform: linux
sys.path: ['/home/stefan/Downloads/urllib3', '/usr/lib/python311.zip', '/usr/lib/python3.11', '/usr/lib/python3.11/lib-dynload', '/home/stefan/Downloads/urllib3/.nox/test-3-11/lib/python3.11/site-packages']
nox > python -bb -m coverage run --parallel-mode -m pytest -r a --color=auto --tb=native --no-success-flaky-report test/
ImportError while loading conftest '/home/stefan/Downloads/urllib3/test/conftest.py'.
test/__init__.py:41: in <module>
    from urllib3 import util
.nox/test-3-11/lib/python3.11/site-packages/urllib3/__init__.py:36: in <module>
    raise ImportError(
E   ImportError: urllib3 v2.0 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with LibreSSL 3.7.2. See: https://github.com/urllib3/urllib3/issues/2168
nox > Command python -bb -m coverage run --parallel-mode -m pytest -r a --color=auto --tb=native --no-success-flaky-report test/ failed with exit code 4
nox > Session test-3.11 failed.
```

**Comment by stefan11111** [comment]
```
~/Downloads/urllib3 $ nox -s test-3.11
nox > Running session test-3.11
nox > Creating virtual environment (virtualenv) using python3.11 in .nox/test-3-11
nox > python -m pip install -r dev-requirements.txt
nox > python -m pip install '.[socks,secure,brotli,zstd]'
nox > pip --version
pip 23.1.2 from /home/stefan/Downloads/urllib3/.nox/test-3-11/lib/python3.11/site-packages/pip (python 3.11)
nox > python --version
Python 3.11.3
nox > python -c 'import struct; print(struct.calcsize('"'"'P'"'"') * 8)'
64
nox > python -m OpenSSL.debug
pyOpenSSL: 23.1.1
cryptography: 39.0.1
cffi: 1.15.1
cryptography's compiled against OpenSSL: OpenSSL 3.0.8 7 Feb 2023
cryptography's linked OpenSSL: OpenSSL 3.0.8 7 Feb 2023
Python's OpenSSL: LibreSSL 3.7.2
Python executable: /home/stefan/Downloads/urllib3/.nox/test-3-11/bin/python
Python version: 3.11.3 (main, May  8 2023, 01:46:20) [GCC 13.1.0]
Platform: linux
sys.path: ['/home/stefan/Downloads/urllib3', '/usr/lib/python311.zip', '/usr/lib/python3.11', '/usr/lib/python3.11/lib-dynload', '/home/stefan/Downloads/urllib3/.nox/test-3-11/lib/python3.11/site-packages']
nox > python -bb -m coverage run --parallel-mode -m pytest --memray --hide-memray-summary -v -ra --color=auto --tb=native --durations=10 --strict-config --strict-markers test/
ERROR: while parsing the following warning configuration:

  default:No IPv6 support. Falling back to IPv4:urllib3.exceptions.HTTPWarning

This error occurred:

Traceback (most recent call last):
  File "/home/stefan/Downloads/urllib3/.nox/test-3-11/lib/python3.11/site-packages/_pytest/config/__init__.py", line 1690, in parse_warning_filter
    category: Type[Warning] = _resolve_warning_category(category_)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/stefan/Downloads/urllib3/.nox/test-3-11/lib/python3.11/site-packages/_pytest/config/__init__.py", line 1728, in _resolve_warning_category
    m = __import__(module, None, None, [klass])
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/stefan/Downloads/urllib3/.nox/test-3-11/lib/python3.11/site-packages/urllib3/__init__.py", line 34, in <module>
    warnings.warn(
urllib3.exceptions.NotOpenSSLWarning: urllib3 v2.0 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 3.7.2'. See: https://github.com/urllib3/urllib3/issues/3020


nox > Command python -bb -m coverage run --parallel-mode -m pytest --memray --hide-memray-summary -v -ra --color=auto --tb=native --durations=10 --strict-config --strict-markers test/ failed with exit code 4
nox > Session test-3.11 failed.
```

**Comment by pquentin** [comment]
Hmm, this is surprising. Can you please try another time after removing lines 25 to 45 in `src/urllib3/__init__.py`? Thanks for the help!

**Comment by stefan11111** [comment]
Here are the results:
```
$ nox -s test-3.11 > test_libressl
nox > Running session test-3.11
nox > Creating virtual environment (virtualenv) using python3.11 in .nox/test-3-11
nox > python -m pip install -r dev-requirements.txt
nox > python -m pip install '.[socks,secure,brotli,zstd]'
nox > pip --version
nox > python --version
nox > python -c 'import struct; print(struct.calcsize('"'"'P'"'"') * 8)'
nox > python -m OpenSSL.debug
nox > python -bb -m coverage run --parallel-mode -m pytest --memray --hide-memray-summary -v -ra --color=auto --tb=native --durations=10 --strict-config --strict-markers test/
nox > Command python -bb -m coverage run --parallel-mode -m pytest --memray --hide-memray-summary -v -ra --color=auto --tb=native --durations=10 --strict-config --strict-markers test/ failed with exit code 1
nox > Session test-3.11 failed.
```
Here is the output:
https://github.com/stefan11111/pastebin/blob/main/test_libressl

Here is the diff of the changes I made:
```
$ git diff
diff --git a/src/urllib3/__init__.py b/src/urllib3/__init__.py
index 7ddfad20..79372db4 100644
--- a/src/urllib3/__init__.py
+++ b/src/urllib3/__init__.py
@@ -22,28 +22,6 @@ from .util.request import make_headers
 from .util.retry import Retry
 from .util.timeout import Timeout

-# Ensure that Python is compiled with OpenSSL 1.1.1+
-# If the 'ssl' module isn't available at all that's
-# fine, we only care if the module is available.
-try:
-    import ssl
-except ImportError:
-    pass
-else:
-    if not ssl.OPENSSL_VERSION.startswith("OpenSSL "):  # Defensive:
-        warnings.warn(
-            "urllib3 v2.0 only supports OpenSSL 1.1.1+, currently "
-            f"the 'ssl' module is compiled with {ssl.OPENSSL_VERSION!r}. "
-            "See: https://github.com/urllib3/urllib3/issues/3020",
-            exceptions.NotOpenSSLWarning,
-        )
-    elif ssl.OPENSSL_VERSION_INFO < (1, 1, 1):  # Defensive:
-        raise ImportError(
-            "urllib3 v2.0 only supports OpenSSL 1.1.1+, currently "
-            f"the 'ssl' module is compiled with {ssl.OPENSSL_VERSION!r}. "
-            "See: https://github.com/urllib3/urllib3/issues/2168"
-        )
-
 # === NOTE TO REPACKAGERS AND VENDORS ===
 # Please delete this block, this logic is only
 # for urllib3 being distributed via PyPI.
```

**Comment by pquentin** [comment]
Awesome, thank you! :100: I'll try analyzing the failures next.

**Comment by stefan11111** [comment]
Is there any cleanup I should do after running those tests?

**Comment by pquentin** [comment]
No, everything is self-contained in the .nox hidden directory.

**Comment by orbea** [comment]
> @stefan11111 @orbea Can I get one of you interested in running the test suite under LibreSSL? I think we have instructions in our docs but essentially it's `pip install nox` then `nox -s test-3.11`.
> 
> Thanks!

Running it with 2.0.2 and LibreSSL 3.7.2 most, but not all of the tests pass.

[urllib3-2.0.2-build.log](https://github.com/urllib3/urllib3/files/11472844/urllib3-2.0.2-build.log)

> On gentoo we do not use pip, instead we install all python packages through portage, the system package manager.

@stefan11111 Fwiw in Gentoo you can run the tests by setting `FEATURES=test`.

**Comment by stefan11111** [comment]
> > @stefan11111 @orbea Can I get one of you interested in running the test suite under LibreSSL? I think we have instructions in our docs but essentially it's `pip install nox` then `nox -s test-3.11`.
> > 
> > Thanks!
> 
> Running it with 2.0.2 and LibreSSL 3.7.2 most, but not all of the tests pass.
> 
> [urllib3-2.0.2-build.log](https://github.com/urllib3/urllib3/files/11472844/urllib3-2.0.2-build.log)
> 
> > On gentoo we do not use pip, instead we install all python packages through portage, the system package manager.
> 
> @stefan11111 Fwiw in Gentoo you can run the tests by setting `FEATURES=test`.

How does this differ form USE="test" ?

**Comment by orbea** [comment]
> How does this differ form USE="test" ?

The Gentoo man pages, wiki and forum should usually explain these questions, for instance.

https://forums.gentoo.org/viewtopic-p-8721179.html

**Comment by pquentin** [comment]
Looking at the failures from @orbea (thank you both!) most of them are fine, it's just that with LibreSSL a different exception is raised than what the tests except:

```
FAILED test/contrib/test_pyopenssl.py::TestClientCerts::test_missing_client_certs_raises_error - OpenSSL.SSL.Error: [('system library', '', 'Connection reset by peer')]
FAILED test/contrib/test_pyopenssl.py::TestSocketClosing::test_socket_close_socket_then_file - OpenSSL.SSL.Error: [('system library', '', 'Bad file descriptor')]
FAILED test/contrib/test_pyopenssl.py::TestSocketSSL::test_ssl_failure_midway_through_conn - AssertionError: Regex pattern did not match.
 Regex: '(wrong version number|record overflow)'
 Input: '("read error: Error([(\'SSL routines\', \'ST_OK\', \'tlsv1 alert protocol version\')])",)'
FAILED test/with_dummyserver/test_socketlevel.py::TestSSL::test_ssl_failure_midway_through_conn - AssertionError: Regex pattern did not match.
 Regex: '(wrong version number|record overflow)'
 Input: '[SSL: TLSV1_ALERT_PROTOCOL_VERSION] tlsv1 alert protocol version (_ssl.c:2588)'
FAILED test/with_dummyserver/test_socketlevel.py::TestProxyManager::test_https_proxymanager_connected_to_http_proxy[http] - assert 'Your proxy appears to only use HTTP and not HTTPS' in "('Unable to connect to proxy', SSLError(SSLError(1, '[SSL: TLSV1_ALERT_PROTOCOL_VERSION] tlsv1 alert protocol version (_ssl.c:1014)')))"
FAILED test/with_dummyserver/test_socketlevel.py::TestProxyManager::test_https_proxymanager_connected_to_http_proxy[https] - assert 'Your proxy appears to only use HTTP and not HTTPS' in "('Unable to connect to proxy', SSLError(SSLError(1, '[SSL: TLSV1_ALERT_PROTOCOL_VERSION] tlsv1 alert protocol version (_ssl.c:1014)')))"
```

However, we have tests that make sure that certificates without Subject Alternative Names (SAN) are rejected, but they appear to be accepted with LibreSSL:

```
FAILED test/with_dummyserver/test_https.py::TestHTTPS_Hostname::test_common_name_without_san_fails[localhost] - Failed: DID NOT RAISE <class 'urllib3.exceptions.MaxRetryError'>
FAILED test/with_dummyserver/test_proxy_poolmanager.py::TestHTTPSProxyVerification::test_https_proxy_no_san[http] - Failed: DID NOT RAISE <class 'urllib3.exceptions.MaxRetryError'>
FAILED test/with_dummyserver/test_proxy_poolmanager.py::TestHTTPSProxyVerification::test_https_proxy_no_san[https] - Failed: DID NOT RAISE <class 'urllib3.exceptions.MaxRetryError'>
```

I'll need to investigate more to understand why.

**Comment by orbea** [comment]
@pquentin In the event these need to be fixed in LibreSSL I would very much appreciate details so that I could report it to them. A simple reproduction would help in that case.

**Comment by pquentin** [comment]
Thanks for the offer. But that shows why supporting LibreSSL is time consuming. We have this failure in Python but someone needs to figure out if this is a CPython bug or LibreSSL bug, we can't just redirect users to the LibreSSL project.

**Comment by orbea** [comment]
Unfortunately I have little specific experience with python and feel out of my depth understanding what the urllib3 tests are doing. However I get your point and I understand why people would be hesitant to have more technical debt, but I also see the benefits of having multiple implementations of crucial software that might expose obscured bugs or support different niches.

**Comment by botovq** [comment]
Hi.

Thanks for taking a relaxation of the library requirements into consideration.

> However, we have tests that make sure that certificates without Subject Alternative Names (SAN) are rejected, but they appear to be accepted with LibreSSL:
> ```
> FAILED test/with_dummyserver/test_https.py::TestHTTPS_Hostname::test_common_name_without_san_fails[localhost] - Failed: DID NOT RAISE <class 'urllib3.exceptions.MaxRetryError'>
> FAILED test/with_dummyserver/test_proxy_poolmanager.py::TestHTTPSProxyVerification::test_https_proxy_no_san[http] - Failed: DID NOT RAISE <class 'urllib3.exceptions.MaxRetryError'>
> FAILED test/with_dummyserver/test_proxy_poolmanager.py::TestHTTPSProxyVerification::test_https_proxy_no_san[https] - Failed: DID NOT RAISE <class 'urllib3.exceptions.MaxRetryError'>
> ```

The reason that this fails is indeed a LibreSSL bug. This is the same bug that is discussed for OpenSSL here:

https://bugs.python.org/issue43522

We are going to have new LibreSSL releases for all stable branches soon (i.e. we will release LibreSSL 3.6.3, 3.7.3, and a dev release 3.8.0 in the coming two weeks) so we could roll a fix for this in. If this is confirmed to make these three tests pass, would that be good enough for a merge or are there more outstanding issues that need to be addressed?

**Comment by sethmlarson** [comment]
@botovq Thank you for responding here with this information! I think if it can be confirmed that our test suite passes using a new build of LibreSSL then that is enough to merge the patch. Is there a place we can follow along on that fix?

**Comment by botovq** [comment]
We do not have an issue for this for you to follow along. As mentioned, I saw one bug in this thread and we will happily fix it. If there are more, please let me know.

Then there is the other class of failures, for which, to quote @pquentin, 

> it's just that with LibreSSL a different exception is raised than what the tests [expect]

Such failures are not a priori LibreSSL bugs. They are equally as likely tests that are written with a particular library in mind. While we may consider addressing some of them over time, if it actually makes sense, this is neither a priority nor are such changes suitable for bundling into a stable release.

**Comment by sethmlarson** [comment]
@botovq No worries on error messages being different, those types of differences are acceptable. I'm more concerned with behavior differences between the different libraries when using the same Python features.

**Comment by botovq** [comment]
For LibreSSL I am using this diff on top of the 2.0.2 tag of this repo to sidestep the test this is about:

```diff
diff --git a/src/urllib3/__init__.py b/src/urllib3/__init__.py
index 43e79fa2..78e73d6f 100644
--- a/src/urllib3/__init__.py
+++ b/src/urllib3/__init__.py
@@ -32,7 +32,7 @@ except ImportError:
 else:
     # fmt: off
     if (
-        not ssl.OPENSSL_VERSION.startswith("OpenSSL ")
+        not ssl.OPENSSL_VERSION.startswith("LibreSSL ")
         or ssl.OPENSSL_VERSION_INFO < (1, 1, 1)
     ):  # Defensive:
         raise ImportError(
```

I tested using LibreSSL 3.7.2 containing a [two-line patch][1] to fix the no-SAN issue pointed out by @pquentin. I got these results:

```
=== 29 failed, 1695 passed, 560 skipped, 148 warnings in 1084.42s (0:18:04) ====
```

When going through the output it turned out that the majority of failing tests were unrelated to LibreSSL. So I ran tests with OpenSSL 3.0.8 on the same machine, same code (minus the diff above), same libs, the only difference being a different Python. Result is this.

```
=== 40 failed, 1684 passed, 560 skipped, 141 warnings in 5392.69s (1:29:52) ====
```

I am not entirely sure where that huge time difference comes from. On that machine (an old i3-based laptop) there is some flakiness to some of the tests as well, with both libs.

One thing I should point out: I used pyca/cryptography linked against OpenSSL 3.0.8. That was not deliberate. I'm only a casual Python user and I could not convince your test tooling to pick up the cryptography I built. I don't think this is very important, though. LibreSSL is well tested with both pyca/cryptography.

At any rate, using packages from a major Linux distribution (Fedora 37), OpenSSL fails to pass a similar number of tests as LibreSSL, so I am not sure we have quality concerns to address. Surely both libraries still have a lot of room for improvement in that department.

Here are the tests that we fail that OpenSSL above doesn't fail:

```
FAILED test/contrib/test_pyopenssl.py::TestSocketSSL::test_ssl_custom_validation_failure_terminates
```

This terminates with:

```
ssl.SSLError: [SSL: SSLV3_ALERT_UNEXPECTED_MESSAGE] sslv3 alert unexpected message (_ssl.c:1007)
```

Not 100% sure where that comes from. It terminates...

```
FAILED test/with_dummyserver/test_socketlevel.py::TestContentFraming::test_framing_set_via_headers[content-length-10-: 10\r\n\r\nxxxxxxxx-lower]
FAILED test/with_dummyserver/test_socketlevel.py::TestContentFraming::test_framing_set_via_headers[content-length-10-: 10\r\n\r\nxxxxxxxx-title]
FAILED test/with_dummyserver/test_socketlevel.py::TestProxyManager::test_https_proxymanager_connected_to_http_proxy[http]
FAILED test/with_dummyserver/test_socketlevel.py::TestProxyManager::test_https_proxymanager_connected_to_http_proxy[https]
```

A bunch of timeouts. Seems to happen in these tests with both libs on this machine occasionally. The network is perfectly good. No idea why this happens.

```
FAILED test/with_dummyserver/test_socketlevel.py::TestSSL::test_requesting_large_resources_via_ssl[False-2147483648]
FAILED test/with_dummyserver/test_socketlevel.py::TestSSL::test_requesting_large_resources_via_ssl[False-None]
FAILED test/with_dummyserver/test_socketlevel.py::TestSSL::test_requesting_large_resources_via_ssl[True-None]
```

These want a wrong version number or a record overflow error, we say bad length.

```
FAILED test/with_dummyserver/test_socketlevel.py::TestSSL::test_ssl_failure_midway_through_conn
```

This test sends plaintext at some point and we raise a wrong version number error from the record layer. This is expected.


[1]: https://marc.info/?l=libressl&m=168457850721143&w=2
[2]: https://gist.github.com/botovq/6d3e14ac385bde71ca9dbd96eb41d142

**Comment by sethmlarson** [comment]
Thanks for running those tests @botovq, not seeing any CN/SAN failures in the list is a good sign. Off-hand you know approximately when that patch will land in LibreSSL and released?

**Comment by botovq** [comment]
On Sun, May 21, 2023 at 08:27:37AM -0700, Seth Michael Larson wrote:
> Thanks for running those tests @botovq, not seeing any CN/SAN failures in the list is a good sign. Off-hand you know approximately when that patch will land in LibreSSL and released?

It won't be long. 1-2 weeks unless something unexpected shows up.

**Comment by pquentin** [comment]
Thanks @botovq, really appreciate your involvement here! I still need to go through your analysis.

@orbea I pushed one commit falling back to our own hostname check for LibreSSL, would you mind trying again?

**Comment by pquentin** [comment]
You're not doing anything wrong, I need to add this line: https://github.com/urllib3/urllib3/pull/3024#issuecomment-1548174223. Is it easy to do on your side for now?

**Comment by pquentin** [comment]
Oh wait, you should test against this pull request, not the main branch.

**Comment by orbea** [comment]
> You're not doing anything wrong, I need to add this line: [#3024 (comment)](https://github.com/urllib3/urllib3/pull/3024#issuecomment-1548174223). Is it easy to do on your side for now?

Thanks for pointing that out, I missed it earlier and yes its easy to apply as a patch.

> Oh wait, you should test against this pull request, not the main branch.

I was using the main branch while applying this PR as a patch, but yes its probably better to just use your branch.

I reran the tests and this is the current status with LibreSSL 3.7.2, I will need to wait for the update for the LibreSSL Portable repo before I can try the fixes on the LibreSSL side.

[build.log](https://github.com/urllib3/urllib3/files/11547989/build.log)

**Comment by botovq** [comment]
On Sun, May 21, 2023 at 08:27:37AM -0700, Seth Michael Larson wrote:
> Thanks for running those tests @botovq, not seeing any CN/SAN failures in the list is a good sign. Off-hand you know approximately when that patch will land in LibreSSL and released?

In today's releases 3.7.3, 3.6.3 on the stable branches (and the dev
release 3.8.0) the bug causing these three CN/SAN test failures is fixed.

**Comment by sethmlarson** [comment]
Thanks @orbea and @botovq for testing out our test suite with the fix to LibreSSL! :rocket: I think from what we know now we can comfortable merge and release this change, what do you think @pquentin and @illia-v?

**Comment by orbea** [comment]
Running the tests again reduced the failures from `44 failed, 1676 passed, 561 skipped, 6 deselected, 149 warnings` with LibreSSL-3.7.2 to `24 failed, 1696 passed, 561 skipped, 6 deselected, 149 warnings` with LibreSSL-3.7.3.

[urllib3-libressl-3.7.3-build.log](https://github.com/urllib3/urllib3/files/11587002/urllib3-libressl-3.7.3-build.log)

@botovq Do any of these look like LibreSSL issues to you?

**Comment by pquentin** [comment]
Thanks @botovq and @orbea! Appreciate the help here.

@sethmlarson I've looked at https://github.com/urllib3/urllib3/pull/3024#issuecomment-1555931127 and the log sent in https://github.com/urllib3/urllib3/pull/3024#issuecomment-1566312427 and I believe we are good to do with this pull request.

**Comment by pquentin** [comment]
The 3.12 failure are unrelated, see https://github.com/urllib3/urllib3/issues/3049. We should add the ast deprecation to filterwarnings.

**Review by sethmlarson — APPROVED** [review]
Excellent, thanks everyone involved for all the work on this. Let's try to get this released on Tuesday :rocket:

**Comment by orbea** [comment]
Thank you very much for relaxing the check, this helps a lot!

File: src/urllib3/exceptions.py [hunk 4]
```diff
@@ -214,6 +214,10 @@ class InsecureRequestWarning(SecurityWarning):
     """Warned when making an unverified HTTPS request."""
 
 
+class NotOpenSSLWarning(SecurityWarning):
+    """Warned when using unsupported SSL library"""
```
Comment (Dsbryant) [inline src/urllib3/exceptions.py:218]: # - [ ]

**Pull request merged**